components/metrics/publisher.go

```diff
@@ -0,0 +1,88 @@
+package metrics
+
+import (
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	publisherLabelKeys = []string{
+		labelKeyHandlerName,
+		labelKeyPublisherName,
+		labelSuccess,
+	}
+)
+
+type PublisherPrometheusMetricsDecorator struct {
+	pub message.Publisher
+
+	publishTimeSeconds  *prometheus.HistogramVec
+	publisherCountTotal prometheus.Gauge
+}
+
+// Publish updates the relevant publisher metrics and calls the wrapped publisher's Publish.
+func (m PublisherPrometheusMetricsDecorator) Publish(topic string, messages ...*message.Message) (err error) {
+	if len(messages) == 0 {
+		return m.pub.Publish(topic)
+	}
+
+	// TODO: take ctx not only from first msg. Might require changing the signature of Publish, which is planned anyway.
+	ctx := messages[0].Context()
+	labels := labelsFromCtx(ctx, publisherLabelKeys...)
+	now := time.Now()
```

Comment:
detail, but `now` is outdated pretty fast, `start` may be better a bit

Reply:
done